In [1]:
import warnings
warnings.filterwarnings("ignore")
import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()
import numpy as np


Instructions for updating:
non-resource variables are not supported in the long term


평균 제곱근 오차(RMSE, Root Mean Squared Error)

임의의 선을 그리고 난 후 이 선이 얼마나 잘 그려졌는지 평가해서 조금씩 수정해나가기 위해 주어진 선의 오차를 평가하는 알고리즘이 필요한데 그 중에서 가장 많이 사용되는 알고리즘이 평균 제곱근 오차이다.

In [2]:
# '04_가중치_바이어스' 예제에서 xData, yData 값이 1건씩 저장된 2차원 리스트
data = [[2, 81], [4, 93], [6, 91], [8, 97]]
# 가중치 a와 바이어스 b를 임의로 정한다. [2.3, 79.0]이 최적의 가중치와 바이어스이다.
ab = [2.3, 79.0]

for i in data:
    print(f'공부 시간: {i[0]}, 공부 시간에 따른 시험 점수: {i[1]}')

공부 시간: 2, 공부 시간에 따른 시험 점수: 81
공부 시간: 4, 공부 시간에 따른 시험 점수: 93
공부 시간: 6, 공부 시간에 따른 시험 점수: 91
공부 시간: 8, 공부 시간에 따른 시험 점수: 97


In [3]:
# 공부 시간과 공부 시간에 따른 시험 점수 데이터 값이 1건씩 저장된 리스트에서 공부 시간과 공부 시간에 따른 시험 점수를 뽑아내서 별도의 리스트에 저장한다.
xData = [d[0] for d in data]
yData = [d[1] for d in data]

for i in range(len(xData)):
    print(f'공부 시간: {xData[i]}, 공부 시간에 따른 시험 점수: {yData[i]}')

공부 시간: 2, 공부 시간에 따른 시험 점수: 81
공부 시간: 4, 공부 시간에 따른 시험 점수: 93
공부 시간: 6, 공부 시간에 따른 시험 점수: 91
공부 시간: 8, 공부 시간에 따른 시험 점수: 97


$Y=ax+b$ 수식에 가중치, 바이어스 값을 대입해서 결과(예측값)를 계산하는 함수

In [4]:
def predict(x):
    return ab[0] * x + ab[1]

In [5]:
# 임의로 정한 가중치와 바이어스로 예측 성적을 계산한다.
predict_result = [] # 예측 점수를 기억할 빈 리스트
for i in range(len(xData)):
    predict_result.append(predict(xData[i]))
    print(f'공부 시간: {xData[i]}, 공부 시간에 따른 시험 점수: {yData[i]}, 예측 점수: {predict_result[i]}')

공부 시간: 2, 공부 시간에 따른 시험 점수: 81, 예측 점수: 83.6
공부 시간: 4, 공부 시간에 따른 시험 점수: 93, 예측 점수: 88.2
공부 시간: 6, 공부 시간에 따른 시험 점수: 91, 예측 점수: 92.8
공부 시간: 8, 공부 시간에 따른 시험 점수: 97, 예측 점수: 97.4


평균 제곱근 오차를 계산하는 함수

In [6]:
def rmse(predict_result, yData):
    # 파이썬의 리스트는 브로드캐스팅 연산이 불가능하므로 넘파이 배열로 바꿔서 연산한다.
    predict_result = np.array(predict_result)
    yData = np.array(yData)
    return np.sqrt(((predict_result - yData) ** 2).mean())

In [7]:
# 평균 제곱근 오차를 계산해서 출력한다.
rmse_result = rmse(predict_result, yData)
print(f'RMSE: {rmse_result}')

RMSE: 2.880972058177584


가중치 3.5, 바이어스 83.0일 경우 RMSE는 10.747092630102339가 계산되는 것을 확인할 수 있다.  
가중치 2.1, 바이어스 76.0일 경우 RMSE는 4.949747468305832가 계산되는 것을 확인할 수 있다.  
이 오차를 줄이면서 새로운 선을 긋는 작업을 해야한다.  
이를 위해서는 가중치와 바이어스를 적절히 조절하면서 오차의 변화를 살펴보고 오차가 최소가되는 가중치와 바이어스를 찾아야 한다.  
앞의 예제에서 계산된 가중치 2.3, 바이어스 79일 경우 RMSE는 2.880972058177584가 계산되는 것을 확인할 수 있다.